### PR TITLE
introduce rich template messaging

### PR DESCRIPTION
**Pull request by coders-kitchen**
and introduce dedicated objects per media type to remove ambiguity when using the API

### CODE REVIEW

File: chat-api/v2/openapi.yaml [hunk 1]
```diff
@@ -1069,25 +1069,83 @@ components:
...
           enum:
             - image
+          example: image
+        url:
+          type: string
+          description: the url of the location where the image is stored
+          example: https://www.tyntec.com/themes/custom/tyntec/image/tyntec-logo-color.png
+        caption:
+          type: string
+          description: additional caption for the image. is shown on the uploaded image when the channel supports this
```
Comment (oneiros-de) [inline chat-api/v2/openapi.yaml]: "It is shown"

File: chat-api/v2/openapi.yaml [hunk 2]
```diff
@@ -1069,25 +1069,83 @@ components:
...
             - document
-            - audio
+          example: document
+        url:
+          type: string
+          description: the url of the location where the document is stored
+          example: https://www.tyntec.com/themes/custom/tyntec/image/tyntec-logo-color.svg
+        caption:
+          type: string
+          description: additional caption for the document. is shown on the uploaded document when the channel supports this
```
Comment (oneiros-de) [inline chat-api/v2/openapi.yaml]: "It is shown"

File: chat-api/v2/openapi.yaml [hunk 3]
```diff
@@ -1069,25 +1069,83 @@ components:
...
+          example: Tyntec Pricelist
+        filename:
+          type: string
+          description: File name of the document to be shown. It's shown on the uploaded document when the cahnnel supports this
```
Comment (oneiros-de) [inline chat-api/v2/openapi.yaml]: "It is shown"

File: chat-api/v2/openapi.yaml [hunk 4]
```diff
@@ -1069,25 +1069,83 @@ components:
...
         url:
           type: string
-          description: the url of the location where the media is stored
+          description: the url of the location where the video is stored
           example: https://www.tyntec.com/themes/custom/tyntec/image/tyntec-logo-color.svg
         caption:
           type: string
-          description: additional caption for the media. is shown on the uploaded media when the channel supports this
-          example: Tyntec Logo
+          description: additional caption for the video. is shown on the uploaded video when the channel supports this
```
Comment (oneiros-de) [inline chat-api/v2/openapi.yaml]: "It is shown"

File: chat-api/v2/openapi.yaml [hunk 5]
```diff
@@ -1101,9 +1159,21 @@ components:
...
+          items:
+            anyOf:
+              - $ref: '#/components/schemas/HeaderComponent'
+              - $ref: '#/components/schemas/BodyComponent'
+        parameters:
+          description: |
+            Parameter for replacement in the template.
+            
+            **NOTE** : This parameter is deprecated by version 2.7 and will be removed in the next major release.
+            Use instead components.
```
Comment (oneiros-de) [inline chat-api/v2/openapi.yaml]: Instead use

File: chat-api/v2/openapi.yaml [hunk 6]
```diff
@@ -1148,11 +1219,83 @@ components:
           - more then 4 consecutive spaces.
 
         It can be formatted accordingly to [Formatting your messages](https://faq.whatsapp.com/en/android/26000002/?fbclid=IwAR1-VtKzKAxleQD-nnWbZpIhLtkNY6Hpjjb4echlEmcfLV2l0C5We4HnsUM).
+        
+        **NOTE** : This parameter is deprecated by version 2.7 and will be removed in the next major release.
+            Use instead components.
```
Comment (oneiros-de) [inline chat-api/v2/openapi.yaml]: Instead use

File: chat-api/v2/openapi.yaml [hunk 7]
```diff
@@ -1069,25 +1069,83 @@ components:
...
+        - $ref: '#/components/schemas/VideoMedia'
+    AudioMedia:
+      type: object
+      description: Audio to be send
       properties:
         type:
           type: string
           description: What kind of media should be send
+          enum:
+            - audio
```
Comment (SpyrosChr) [inline chat-api/v2/openapi.yaml]: - audio

and then next line 

example: image

File: chat-api/v2/openapi.yaml [hunk 8]
```diff
@@ -1069,25 +1069,83 @@ components:
...
+        - $ref: '#/components/schemas/DocumentMedia'
+        - $ref: '#/components/schemas/VideoMedia'
+    AudioMedia:
+      type: object
+      description: Audio to be send
       properties:
         type:
           type: string
           description: What kind of media should be send
+          enum:
```
Comment (SpyrosChr) [inline chat-api/v2/openapi.yaml]: maybe this is overkill. This is only audio, why do you need enum for only 1 option

File: chat-api/v2/openapi.yaml [hunk 9]
```diff
@@ -1069,25 +1069,83 @@ components:
...
+        - $ref: '#/components/schemas/ImageMedia'
+        - $ref: '#/components/schemas/DocumentMedia'
+        - $ref: '#/components/schemas/VideoMedia'
+    AudioMedia:
+      type: object
+      description: Audio to be send
       properties:
         type:
           type: string
           description: What kind of media should be send
```
Comment (SpyrosChr) [inline chat-api/v2/openapi.yaml]: a little bit missleading since we are talking about only audio

File: chat-api/v2/openapi.yaml [hunk 10]
```diff
@@ -1069,25 +1069,83 @@ components:
...
+          description: the url of the location where the audio file is stored
+          example: https://www.tyntec.com/themes/custom/tyntec/image/tyntec-logo-color.svg
+    ImageMedia:
+      type: object
+      description: Image to be send
+      properties:
+        type:
+          type: string
+          description: What kind of media should be send. Can only be png or jpeg
           enum:
```
Comment (SpyrosChr) [inline chat-api/v2/openapi.yaml]: again this is a little bit overkill. enum with a single option

File: chat-api/v2/openapi.yaml [hunk 11]
```diff
@@ -1069,25 +1069,83 @@ components:
...
+          example: https://www.tyntec.com/themes/custom/tyntec/image/tyntec-logo-color.png
+        caption:
+          type: string
+          description: additional caption for the image. It is shown on the uploaded image when the channel supports this
+          example: Tyntec Logo
+    DocumentMedia:
+      type: object
+      description: Document to be send
+      properties:
+        type:
```
Comment (SpyrosChr) [inline chat-api/v2/openapi.yaml]: same as above

File: chat-api/v2/openapi.yaml [hunk 12]
```diff
@@ -1069,25 +1069,83 @@ components:
...
+          example: Tyntec Pricelist
+        filename:
+          type: string
+          description: File name of the document to be shown. It is shown on the uploaded document when the channel supports this
+          example: tyntec_pricelist.pdf
+    VideoMedia:
+      type: object
+      description: Video to be send
+      properties:
+        type:
```
Comment (SpyrosChr) [inline chat-api/v2/openapi.yaml]: same as above

File: chat-api/v2/openapi.yaml [hunk 13]
```diff
@@ -1069,25 +1069,83 @@ components:
...
+          type: string
+          description: What kind of media should be send
+          enum:
             - video
```
Comment (SpyrosChr) [inline chat-api/v2/openapi.yaml]: video with image example below

**Comment by SpyrosChr** [comment]
Generally i see that there are many enums with only one option. I don't know why you did it that way. maybe for a reason. I just noticed that.

**Comment by coders-kitchen** [comment]
Yes. Only option to express a constant value with open api specification.

File: chat-api/v2/openapi.yaml (hunk 7 above)
Comment (coders-kitchen) [reply chat-api/v2/openapi.yaml]: fixed

File: chat-api/v2/openapi.yaml (hunk 8 above)
Comment (coders-kitchen) [reply chat-api/v2/openapi.yaml]: see comment in the main section :)

File: chat-api/v2/openapi.yaml (hunk 9 above)
Comment (coders-kitchen) [reply chat-api/v2/openapi.yaml]: fixed

File: chat-api/v2/openapi.yaml (hunk 10 above)
Comment (coders-kitchen) [reply chat-api/v2/openapi.yaml]: see comment in the main section :)

File: chat-api/v2/openapi.yaml (hunk 11 above)
Comment (coders-kitchen) [reply chat-api/v2/openapi.yaml]: see comment in the main section :)

File: chat-api/v2/openapi.yaml (hunk 12 above)
Comment (coders-kitchen) [reply chat-api/v2/openapi.yaml]: see comment in the main section :)

File: chat-api/v2/openapi.yaml (hunk 13 above)
Comment (coders-kitchen) [reply chat-api/v2/openapi.yaml]: see comment in the main section :)

**Comment by coders-kitchen** [comment]
Additionally it's the only way to safely express the differences on the additional properties on the supported media types.

**Pull request merged**